astropy/cosmology/connect.py

```diff
@@ -0,0 +1,218 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+import copy
+import warnings
+
+from astropy.io import registry as io_registry
+from astropy.utils.exceptions import AstropyUserWarning
+
+__all__ = ["CosmologyRead", "CosmologyWrite"]
+__doctest_skip__ = ["CosmologyRead", "CosmologyWrite"]
+
+
+class CosmologyRead(io_registry.UnifiedReadWrite):
+    """Read and parse data to a `~astropy.cosmology.Cosmology`.
+
+    This can *only* be called from `~astropy.cosmology.Cosmology`,
+    not any subclasses.
+
+    This function provides the Cosmology interface to the Astropy unified I/O
+    layer. This allows easily reading a file in supported data formats using
+    syntax such as::
+
+        >>> from astropy.cosmology import Cosmology
+        >>> cosmo1 = Cosmology.read('cosmo1.ecsv')
+        >>> cosmo2 = Cosmology.read('cosmo2.json')
+
+    Get help on the available readers using the ``help()`` method::
+
+      >>> Cosmology.read.help()  # Get help reading and list supported formats
+      >>> Cosmology.read.help('json')  # Get detailed help on JSON reader
+      >>> Cosmology.read.list_formats()  # Print list of available formats
+
+    See also: https://docs.astropy.org/en/stable/io/unified.html
+
+    Parameters
+    ----------
+    *args : (optional)
+        Positional arguments passed through to data reader. If supplied the
+        first argument is typically the input filename.
+    format : str (optional, keyword-only)
+        File format specifier.
+    **kwargs
+        Keyword arguments passed through to data reader.
+
+    Returns
+    -------
+    out : `~astropy.cosmology.Cosmology` subclass instance
+        `~astropy.cosmology.Cosmology` corresponding to file contents.
+
+    Methods
+    -------
+    from_mapping
+        Parse a mapping instance (e.g. dict) into a cosmology instance.
+    from_table
+        Parse a `~astropy.table.Table` instance into a cosmology instance.
+
+    Notes
+    -----
+    The cosmology class must be specified by name in the metadata (if a |Table|)
+    or as a field titled ``cosmology`` if a mapping.
+    Metadata can be included on the |Table| or as a key "meta" for a JSON.
+
+    Warns
+    -----
+    `~astropy.utils.exceptions.AstropyUserWarning`
+        If ``read`` is examined not from the Cosmology base class.
+    """
+
+    def __new__(cls, instance, cosmo_cls):
+        from astropy.cosmology.core import Cosmology
+
+        # warn that ``read`` can only be called from the base class.
+        # and return None to prevent usage.
+        if cosmo_cls is not Cosmology:
+            warnings.warn(("``Cosmology.read()`` must be called from the "
```

Comment:
In my first imagining of Cosmo I/O, that was the implementation. The logic got a little complicated, so it was suggested to simplify. I agree that it would be nice for subclasses to implement ``.read``, checking for consistency and providing defaults. Were we to remove ``__new__`` before implementing those checks / defaults, I think the behavior of ``SubClass.read()`` could reasonably be interpreted as a bug. I propose: changing the docs to say that  ``SubClass.read()`` is not yet implemented, returning ``NotImplemented`` instead of None for ``SubClass.read``, and making an Issue for followup where the consistency checks and providing-of-defaults is all implemented.

Reply:
> Why is this a warning and not an exception? After this warning then the user immediately gets an obscure exception and might panic and not even notice the warning above.

I flip-flopped a few times on this implementation. Ultimately it felt the lesser of 3 evils.
1) Merely typing ``SubClass.read`` raises an Exception. I don't think any attribute/descriptor should ever raise an error merely by typing it.
2) It returns a no-op — e.g. all methods like ``SubClass.read.__call__`` will raise an error if used.
3) ``SubClass.read`` returns something that doesn't raise an Exception and obviously cannot be used for reading. I do agree that "None" is probably not the best thing to return. With the planned implementation of reading from a subclass (above 2 comments), I think returning "NotImplemented" is more clear and a reasonable stop-gap. Now the error is

```
In [6]: Planck18.read('table.ecsv')
WARNING: ``Cosmology.read()`` is not (yet) implemented for ``Cosmology`` subclasses. [astropy.cosmology.connect]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-be56557be2c8> in <module>
----> 1 Planck18.read('table.ecsv')

TypeError: 'NotImplemented' object is not callable
```